1. Create a simple Keras Sequential model for classifying handwritten digits (use the MNIST dataset), and train it without any regularization. Save the validation accuracy after 5 epochs.

In [20]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten

In [2]:
(X_train , y_train) , (X_test , y_test) = tf.keras.datasets.mnist.load_data()

In [3]:
X_train = X_train/255.0
x_test = X_test/255.

In [6]:
model = Sequential([
    Flatten(input_shape = (28,28)),
    Dense(128, activation = 'relu'),
    Dense(10, activation='softmax')
])

In [7]:
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])

In [8]:
history = model.fit(X_train , y_train , epochs = 5, validation_data = (X_test,y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 19s 8ms/step - accuracy: 0.9261 - loss: 0.2601 - val_accuracy: 0.9570 - val_loss: 18.8198
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9658 - loss: 0.1162 - val_accuracy: 0.9715 - val_loss: 13.3557
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9764 - loss: 0.0789 - val_accuracy: 0.9710 - val_loss: 13.9178
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9819 - loss: 0.0598 - val_accuracy: 0.9742 - val_loss: 13.8883
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9856 - loss: 0.0449 - val_accuracy: 0.9770 - val_loss: 14.2086


In [17]:
val_accuracy = history.history['val_accuracy']

In [18]:
print("Validation Accuracy after 5 epochs:",val_accuracy)

Validation Accuracy after 5 epochs: [0.9570000171661377, 0.9714999794960022, 0.9710000157356262, 0.9742000102996826, 0.9769999980926514]


2. Modify your Keras model to add L2 regularization (kernel_regularizer=regularizers.l2(0.01)) to all Dense layers, retrain for 5 epochs, and compare the validation accuracy to your previous result

In [22]:
from tensorflow.keras import regularizers

In [25]:
model = Sequential([
    Flatten(input_shape = (28,28)),
    Dense(128,activation="relu",kernel_regularizer= regularizers.l2(0.01)),
    Dense(10, activation='softmax', kernel_regularizer = regularizers.l2(0.01))
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [27]:
model.compile(optimizer='adam', loss = 'sparse_categorical_crossentropy',
              metrics =['accuracy'])

In [29]:
history = model.fit(X_train, y_train, epochs = 5, validation_data = (x_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9017 - loss: 0.7349 - val_accuracy: 0.9108 - val_loss: 0.7090
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9009 - loss: 0.7304 - val_accuracy: 0.9092 - val_loss: 0.7058
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9028 - loss: 0.7266 - val_accuracy: 0.9071 - val_loss: 0.7123
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9032 - loss: 0.7237 - val_accuracy: 0.9101 - val_loss: 0.7059
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9043 - loss: 0.7217 - val_accuracy: 0.9074 - val_loss: 0.7064


In [34]:
val_accuracy = history.history['val_accuracy']

In [35]:
print("Validation Accuracy after 5 epochs:",val_accuracy)

Validation Accuracy after 5 epochs: [0.9107999801635742, 0.9092000126838684, 0.9071000218391418, 0.910099983215332, 0.9074000120162964]


| Model | Validation Accuracy (Epoch 1 → 5) | Observation |
|-------|------------------------------------|-------------|
| **Without Regularization** | **[0.9570, 0.9715, 0.9710, 0.9742, 0.9770]** | Validation accuracy steadily improves over the 5 epochs, reaching **97.70%**. The model learns quickly but may be more prone to overfitting. |
| **With L2 Regularization (`regularizers.l2(0.01)`)** | **[0.9108, 0.9092, 0.9071, 0.9101, 0.9074]** | Validation accuracy remains around **91%** throughout training. The strong L2 penalty limits the model's complexity, resulting in lower accuracy but reducing the risk of overfitting. |

3. Add Dropout layers (with rate=0.5) after each Dense layer in your Keras model, retrain for 5 epochs, and observe the change in validation accuracy.<br><br><em><strong>Hint:</strong> Use keras.layers.Dropout and place it after activations.</em>

In [36]:
from tensorflow.keras.layers import Dropout

In [38]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation = 'relu'),
    Dropout(0.5),
    Dense(10,activation = 'softmax'),
    Dropout(0.5)
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [39]:
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics =['accuracy'])

In [40]:
history = model.fit(X_train , y_train , epochs = 5 , validation_data= (X_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 8ms/step - accuracy: 0.4391 - loss: 3.3985 - val_accuracy: 0.9024 - val_loss: 284.4554
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.4600 - loss: 2.3135 - val_accuracy: 0.9180 - val_loss: 301.0623
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.4730 - loss: 2.1007 - val_accuracy: 0.9229 - val_loss: 317.5916
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.4748 - loss: 1.9890 - val_accuracy: 0.9319 - val_loss: 315.0295
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.4786 - loss: 1.9366 - val_accuracy: 0.9331 - val_loss: 329.1656


In [41]:
val_accuracy = history.history['val_accuracy']

In [42]:
print('Validation Accuracy after 5 epochs and adding dropout:', val_accuracy)

Validation Accuracy after 5 epochs and adding dropout: [0.902400016784668, 0.9179999828338623, 0.9229000210762024, 0.9319000244140625, 0.9330999851226807]


4. Integrate BatchNormalization layers into your Keras model (before activation functions in Dense layers), retrain, and note how the training and validation accuracy curves change.<br><br><em><strong>Hint:</strong> Use keras.layers.BatchNormalization before the activation in each Dense block.</em>

In [44]:
from tensorflow.keras.layers import BatchNormalization, Activation

In [46]:
model = Sequential([
    Flatten(input_shape=(28,28)),

    Dense(128),
    BatchNormalization(),
    Activation('relu'),

    Dense(10),
    BatchNormalization(),
    Activation('softmax')
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [47]:
model.compile(optimizer = 'adam', loss ='sparse_categorical_crossentropy', metrics =['accuracy'])

In [48]:
history = model.fit(X_train , y_train , epochs = 5 , validation_data = (X_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 24s 8ms/step - accuracy: 0.9093 - loss: 0.4461 - val_accuracy: 0.8411 - val_loss: 48.3958
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.9514 - loss: 0.2121 - val_accuracy: 0.8080 - val_loss: 71.2999
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9637 - loss: 0.1511 - val_accuracy: 0.8374 - val_loss: 64.7332
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9691 - loss: 0.1196 - val_accuracy: 0.8184 - val_loss: 96.9460
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9745 - loss: 0.0996 - val_accuracy: 0.8112 - val_loss: 108.4668


In [49]:
print('Training Accuracy:', history.history['accuracy'])
print('Validation Accuracy:', history.history['val_accuracy'])

Training Accuracy: [0.9092666506767273, 0.9514333605766296, 0.9636666774749756, 0.9690999984741211, 0.9744666814804077]
Validation Accuracy: [0.8410999774932861, 0.8080000281333923, 0.8374000191688538, 0.8184000253677368, 0.8112000226974487]


| Model | Training Accuracy (Epoch 1 → 5) | Validation Accuracy (Epoch 1 → 5) | Observation |
|-------|----------------------------------|------------------------------------|-------------|
| With Batch Normalization | **[0.9093, 0.9514, 0.9637, 0.9691, 0.9745]** | **[0.8411, 0.8080, 0.8374, 0.8184, 0.8112]** | The training accuracy increased steadily, but the validation accuracy remained much lower and even decreased over time. This indicates that the model is not generalizing well to the validation data and may be overfitting or using an unsuitable Batch Normalization configuration. |

5. Implement EarlyStopping in your model training using keras.callbacks.EarlyStopping with patience=2, and show how it affects the number of training epochs and final validation accuracy.

In [50]:
from tensorflow.keras.callbacks import EarlyStopping

In [51]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128,activation='relu'),
    Dense(10,activation='softmax')
])

c:\Users\hp\.conda\envs\tfenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [52]:
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics =['accuracy'])

In [53]:
early_stopping = EarlyStopping(monitor= 'val_loss', patience = 2,restore_best_weights = True)

In [54]:
history = model.fit(
    X_train , y_train ,
    epochs = 5 ,
    validation_data =(X_test, y_test),
    callbacks=[early_stopping]
    )

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 16s 7ms/step - accuracy: 0.9260 - loss: 0.2621 - val_accuracy: 0.9567 - val_loss: 18.1297
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9660 - loss: 0.1150 - val_accuracy: 0.9693 - val_loss: 14.2914
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9762 - loss: 0.0790 - val_accuracy: 0.9729 - val_loss: 14.8104
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9818 - loss: 0.0594 - val_accuracy: 0.9765 - val_loss: 13.1425
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9859 - loss: 0.0462 - val_accuracy: 0.9732 - val_loss: 14.8739


In [55]:
print("Number of epochs trained:", len(history.history['loss']))
print("Final Validation Accuracy:", history.history['val_accuracy'][-1])

Number of epochs trained: 5
Final Validation Accuracy: 0.9732000231742859


| Without Early Stopping | With Early Stopping (patience=2) |
|-------------------------|----------------------------------|
| Trains for all specified epochs (e.g., 20), even if validation performance stops improving. | Stops training automatically when the validation loss does not improve for 2 consecutive epochs. |
| May lead to overfitting and unnecessary training time. | Reduces overfitting, saves training time, and restores the best model weights. |
| Final validation accuracy depends on the last epoch. | Final validation accuracy corresponds to the best-performing model before training stopped. |